src/Evolution/DiscontinuousGalerkin/SlopeLimiters/Hierarchical.hpp

```diff
@@ -0,0 +1,982 @@
+      gsl::not_null<bool*> limited_any_component, const Mesh<2>& mesh,
+      const std::unordered_map<
+          std::pair<Direction<2>, ElementId<2>>, PackagedData,
+          boost::hash<std::pair<Direction<2>, ElementId<2>>>>& neighbor_data)
+      const noexcept;
+  template <typename Tag>
+  char limit_one_tensor(
+      gsl::not_null<Variables<tmpl::list<ModalTag<Tags>...>>*> coeffs_self,
+      gsl::not_null<bool*> limited_any_component, const Mesh<3>& mesh,
+      const std::unordered_map<
+          std::pair<Direction<3>, ElementId<3>>, PackagedData,
+          boost::hash<std::pair<Direction<3>, ElementId<3>>>>& neighbor_data)
+      const noexcept;
+
+  SPECTRE_ALWAYS_INLINE static size_t storage_index(const Mesh<2>& mesh,
+                                                    const size_t i,
+                                                    const size_t j) noexcept {
+    ASSERT(i < mesh.extents()[0], "first coefficient index must be less than "
+                                      << mesh.extents()[0] << " but is " << i);
+    ASSERT(j < mesh.extents()[1], "second coefficient index must be less than "
+                                      << mesh.extents()[1] << " but is " << j);
+    return i + mesh.extents()[0] * j;
+  }
+
+  SPECTRE_ALWAYS_INLINE static size_t storage_index(const Mesh<3>& mesh,
+                                                    const size_t i,
+                                                    const size_t j,
+                                                    const size_t k) noexcept {
+    ASSERT(i < mesh.extents()[0], "first coefficient index must be less than "
+                                      << mesh.extents()[0] << " but is " << i);
+    ASSERT(j < mesh.extents()[1], "second coefficient index must be less than "
+                                      << mesh.extents()[1] << " but is " << j);
+    ASSERT(j < mesh.extents()[2], "third coefficient index must be less than "
+                                      << mesh.extents()[2] << " but is " << k);
+    return i + mesh.extents()[0] * (j + mesh.extents()[1] * k);
+  }
+
+  template <typename Tag, size_t Dim>
+  char fill_variables_tag_with_spectral_coeffs(
+      const gsl::not_null<Variables<tmpl::list<ModalTag<Tags>...>>*>
+          modal_coeffs,
+      const db::item_type<Tag>& nodal_tensor, const Mesh<Dim>& mesh) const
+      noexcept;
+
+  HierarchicalType hierarchical_type_{};
+  std::vector<double> alphas_{};
+  bool disable_for_debugging_{false};
+};
+
+template <size_t VolumeDim, typename... Tags>
+Hierarchical<VolumeDim, tmpl::list<Tags...>>::Hierarchical(
+    HierarchicalType hierarchical_type, std::vector<double> alphas,
+    bool disable_for_debugging, const OptionContext& contex)
+    : hierarchical_type_(hierarchical_type),
+      alphas_(std::move(alphas)),
+      disable_for_debugging_(disable_for_debugging) {
+  if (alg::any_of(alphas_, [](const double t) noexcept {
+        return t > 1.0 or t <= 0.0;
```

Comment:
Because her bounds are stricter than necessary. The lower bound is derived as the "exact" value that "should" be chosen, but being larger or smaller makes the limiter less or more dissipative, respectively.

Reply:
Okay, can you make a note in the documentation that the alpha is only checked to lie in [0,1] which is a looser interval than the one recommended in the paper?